libvast/src/pipeline.cpp

```diff
@@ -0,0 +1,187 @@
+                                         repr));
+    }
+    // 2. find plugin using operator name
+    const auto* plugin = plugins::find<operator_plugin>(plugin_name);
+    if (!plugin) {
+      return caf::make_error(ec::syntax_error,
+                             fmt::format("failed to parse pipeline '{}': "
+                                         "operator '{}' does not exist",
+                                         repr, plugin_name));
+    }
+    // 3. ask the plugin to parse itself from the remainder
+    auto [remaining_repr, op] = plugin->make_operator(std::string_view{f, l});
+    if (!op)
+      return caf::make_error(ec::unspecified,
+                             fmt::format("failed to parse pipeline '{}': {}",
+                                         repr, op.error()));
+    ops.push_back(std::move(*op));
+    repr = remaining_repr;
+  }
+  return pipeline{std::move(ops)};
+}
+
+auto pipeline::copy() const -> operator_ptr {
+  auto copied = std::make_unique<pipeline>();
+  copied->operators_.reserve(operators_.size());
+  for (auto& op : operators_) {
+    copied->operators_.push_back(op->copy());
+  }
+  return copied;
+};
+
+auto pipeline::to_string() const -> std::string {
+  return fmt::to_string(fmt::join(operators_, " | "));
+}
+
+auto pipeline::instantiate(operator_input input,
+                           operator_control_plane& control) const
+  -> caf::expected<operator_output> {
+  if (operators_.empty()) {
+    auto f = detail::overload{
+      [](std::monostate) -> operator_output {
+        return generator<std::monostate>{};
+      },
+      []<class Input>(generator<Input> input) -> operator_output {
+        return input;
+      },
+    };
+    return std::visit(f, std::move(input));
+  }
+  auto it = operators_.begin();
+  auto end = operators_.end();
+  while (true) {
+    auto output = (*it)->instantiate(std::move(input), control);
+    if (!output) {
+      return output.error();
+    }
+    ++it;
+    if (it == end) {
+      return output;
+    }
+    auto f = detail::overload{
+      [](generator<std::monostate>) -> operator_input {
+        return std::monostate{};
+      },
+      []<class Output>(generator<Output> output) -> operator_input {
+        return output;
+      },
+    };
+    input = std::visit(f, std::move(*output));
+    if (std::holds_alternative<std::monostate>(input)) {
+      return caf::make_error(ec::type_clash, "pipeline ended before all "
+                                             "operators were used");
+    }
+  }
+}
+
+auto make_local_executor(pipeline p) -> generator<caf::expected<void>> {
+  local_control_plane ctrl;
+  auto dynamic_gen = p.instantiate(std::monostate{}, ctrl);
+  if (!dynamic_gen) {
+    co_yield std::move(dynamic_gen.error());
+    co_return;
+  }
+  auto gen = std::get_if<generator<std::monostate>>(&*dynamic_gen);
+  if (!gen) {
+    co_yield caf::make_error(ec::logic_error,
+                             "right side of pipeline is not closed");
```

Comment:
This new API allows for calling `pipeline::instantiate` twice, but is that safe to do?

If yes, then `make_local_executor` should take the pipeline by const-ref.

Reply:
Yes, and I think that is a feature.

We could use const-ref for `make_local_executor`, but that means that you have to keep the pipeline object alive (so probably at least const-pointer to make this obvious). Or we keep it as-is for now, we can always relax this restriction later.

Reply:
Should make_local_executor then be a member function of `pipeline`?

Reply:
Ah nvm. Let's just resolve this, I don't think it's worth spending more time on this.